File: collapsingtoolbarlayout-subtitle/tests/src/com/hendraanggrian/collapsingtoolbarlayout/subtitle/test/InstrumentedTest.java
package com.hendraanggrian.collapsingtoolbarlayout.subtitle.test;

import android.support.annotation.NonNull;
import android.support.design.widget.SubtitleCollapsingToolbarLayout;
import android.support.test.espresso.ViewAction;
import android.support.test.espresso.action.CoordinatesProvider;
import android.support.test.espresso.action.GeneralLocation;
import android.support.test.espresso.action.GeneralSwipeAction;
import android.support.test.espresso.action.Press;
import android.support.test.espresso.action.Swipe;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.view.Gravity;
import android.view.View;

import com.hendraanggrian.widget.ErrorView;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.actionWithAssertions;
import static android.support.test.espresso.action.ViewActions.swipeDown;
import static android.support.test.espresso.matcher.ViewMatchers.withId;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
@RunWith(AndroidJUnit4.class)
public class InstrumentedTest {

    private static final float EDGE_FUZZ_FACTOR = 0.083f;

    @Rule
    public ActivityTestRule<InstrumentedActivity> rule = new ActivityTestRule<>(InstrumentedActivity.class);

    @Test
    public void gravity() {
        turn(new OnPerform<SubtitleCollapsingToolbarLayout>() {
            @Override
            public void onPerform(SubtitleCollapsingToolbarLayout layout) {
            }
        });
        turn(new OnPerform<SubtitleCollapsingToolbarLayout>() {
            @Override
            public void onPerform(SubtitleCollapsingToolbarLayout layout) {
                layout.setExpandedTitleGravity(Gravity.END);
                layout.setCollapsedTitleGravity(Gravity.CENTER);
            }
        });
    }

    public static void turn(final OnPerform<SubtitleCollapsingToolbarLayout> perform) {
        onView(withId(R.id.errorView)).perform(new SimpleViewAction<ErrorView>(ErrorView.class) {
            @Override
            public void onPerform(ErrorView view) {
                view.setLogoDrawable(R.drawable.up);
                view.setText("Swiping up...");
            }
        });
        onView(withId(R.id.toolbarLayout)).perform(new SimpleViewAction<SubtitleCollapsingToolbarLayout>(SubtitleCollapsingToolbarLayout.class) {
            @Override
            public void onPerform(SubtitleCollapsingToolbarLayout layout) {
                perform.onPerform(layout);
            }
        }, slowerSwipeUp());
        onView(withId(R.id.errorView)).perform(new SimpleViewAction<ErrorView>(ErrorView.class) {
            @Override
            public void onPerform(ErrorView view) {
                view.setLogoDrawable(R.drawable.down);
                view.setText("Swiping down...");
            }
        });
        onView(withId(R.id.toolbar)).perform(swipeDown(), swipeDown(), swipeDown(), swipeDown(), swipeDown());
        onView(withId(R.id.errorView)).perform(new SimpleViewAction<ErrorView>(ErrorView.class) {
            @Override
            public void onPerform(ErrorView view) {
                view.setLogoDrawable(null);
                view.setText("Done");
            }
        });
    }

    @NonNull
    public static ViewAction slowerSwipeUp() {
        return actionWithAssertions(new GeneralSwipeAction(SlowerSwipe.INSTANCE,
                translate(GeneralLocation.BOTTOM_CENTER, 0, -EDGE_FUZZ_FACTOR),
                GeneralLocation.TOP_CENTER, Press.FINGER));
    }

    @NonNull
    public static ViewAction slowerSwipeDown() {
        return actionWithAssertions(new GeneralSwipeAction(Swipe.SLOW,
                translate(GeneralLocation.TOP_CENTER, 0, EDGE_FUZZ_FACTOR),
                GeneralLocation.BOTTOM_CENTER, Press.FINGER));
    }

    @NonNull
    private static CoordinatesProvider translate(final CoordinatesProvider coords, final float dx, final float dy) {
        return new CoordinatesProvider() {
            @Override
            public float[] calculateCoordinates(View view) {
                float xy[] = coords.calculateCoordinates(view);
                xy[0] += dx * view.getWidth();
                xy[1] += dy * view.getHeight();
                return xy;
            }
        };
    }
}

File: collapsingtoolbarlayout-subtitle/tests/src/com/hendraanggrian/collapsingtoolbarlayout/subtitle/test/InstrumentedTest.kt
package com.hendraanggrian.collapsingtoolbarlayout.subtitle.test

import android.graphics.Typeface
import android.support.design.widget.Errorbar
import android.support.design.widget.SubtitleCollapsingToolbarLayout
import android.support.design.widget.errorbar
import android.support.test.InstrumentationRegistry.getTargetContext
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.*
import android.support.test.espresso.action.ViewActions.actionWithAssertions
import android.support.test.espresso.action.ViewActions.swipeDown
import android.support.test.espresso.matcher.ViewMatchers.withId
import android.support.test.filters.LargeTest
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.support.v7.widget.Toolbar
import android.view.Gravity
import android.widget.FrameLayout
import com.hendraanggrian.collapsingtoolbarlayout.subtitle.test.activity.InstrumentedActivity
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
@LargeTest
class InstrumentedTest {

    @Rule @JvmField val rule = ActivityTestRule(InstrumentedActivity::class.java)
    lateinit var errorbar: Errorbar

    @Before
    fun errorbarInit() {
        onView(withId(R.id.frameLayout)).perform(object : SimpleViewAction<FrameLayout>(FrameLayout::class.java) {
            override fun onPerform(t: FrameLayout) {
                errorbar = errorbar(t, "")
            }
        })
    }

    @Test
    @Throws(Exception::class)
    fun gravity() {
        turn({ })
        turn({ t ->
            t.expandedTitleGravity = Gravity.END
            t.collapsedTitleGravity = Gravity.CENTER
        })
    }

    @Test
    @Throws(Exception::class)
    fun typeface() {
        turn({ })
        turn({ t ->
            val assets = getTargetContext().assets
            t.expandedTitleTypeface = Typeface.createFromAsset(assets, "SourceCodePro-Bold.ttf")
            t.expandedSubtitleTypeface = Typeface.createFromAsset(assets, "SourceCodePro-Regular.ttf")
        })
    }

    private fun turn(perform: (SubtitleCollapsingToolbarLayout) -> Unit) {
        onView(withId(R.id.toolbarLayout)).perform(
                object : SimpleViewAction<SubtitleCollapsingToolbarLayout>(SubtitleCollapsingToolbarLayout::class.java) {
                    override fun onPerform(t: SubtitleCollapsingToolbarLayout) {
                        perform.invoke(t)
                        errorbar.setLogoResource(R.drawable.up)
                        errorbar.setText("Swiping up...")
                    }
                },
                slowerSwipeUp())
        onView(withId(R.id.toolbar)).perform(
                object : SimpleViewAction<Toolbar>(Toolbar::class.java) {
                    override fun onPerform(t: Toolbar) {
                        errorbar.setLogoResource(R.drawable.down)
                        errorbar.setText("Swiping down...")
                    }
                },
                swipeDown(),
                swipeDown(),
                swipeDown(),
                swipeDown(),
                swipeDown(),
                object : SimpleViewAction<Toolbar>(Toolbar::class.java) {
                    override fun onPerform(t: Toolbar) {
                        errorbar.setLogoDrawable(null)
                        errorbar.setText("Done")
                    }
                })
    }

    companion object {
        private val EDGE_FUZZ_FACTOR = 0.083f

        private fun slowerSwipeUp() = actionWithAssertions(GeneralSwipeAction(SlowerSwipe.INSTANCE,
                translate(GeneralLocation.BOTTOM_CENTER, 0f, -EDGE_FUZZ_FACTOR),
                GeneralLocation.TOP_CENTER, Press.FINGER))

        private fun slowerSwipeDown() = actionWithAssertions(GeneralSwipeAction(Swipe.SLOW,
                translate(GeneralLocation.TOP_CENTER, 0f, EDGE_FUZZ_FACTOR),
                GeneralLocation.BOTTOM_CENTER, Press.FINGER))

        private fun translate(coords: CoordinatesProvider, dx: Float, dy: Float) = CoordinatesProvider { view ->
            val xy = coords.calculateCoordinates(view)
            xy[0] += dx * view.width
            xy[1] += dy * view.height
            xy
        }
    }
}